src/core.js

```diff
@@ -67,36 +67,36 @@ extend( QUnit, {
 	on,
 
 	// Call on start of module test to prepend name to all tests
-	module: function( name, testEnvironment, executeNow ) {
+	module( name, options, executeNow ) {
 		if ( arguments.length === 2 ) {
-			if ( objectType( testEnvironment ) === "function" ) {
-				executeNow = testEnvironment;
-				testEnvironment = undefined;
+			if ( objectType( options ) === "function" ) {
+				executeNow = options;
+				options = undefined;
 			}
 		}
 
-		let module = createModule( name, testEnvironment );
+		let module = createModule( name, options );
 
 		// Move any hooks to a 'hooks' object
-		if ( module.testEnvironment ) {
-			module.hooks = {
-				before: module.testEnvironment.before,
-				beforeEach: module.testEnvironment.beforeEach,
-				afterEach: module.testEnvironment.afterEach,
-				after: module.testEnvironment.after
-			};
-
-			delete module.testEnvironment.before;
-			delete module.testEnvironment.beforeEach;
-			delete module.testEnvironment.afterEach;
-			delete module.testEnvironment.after;
+		const testEnvironment = module.testEnvironment;
+		const hooks = module.hooks = {};
+
+		setHookFromEnvironment( hooks, testEnvironment, "before" );
+		setHookFromEnvironment( hooks, testEnvironment, "beforeEach" );
+		setHookFromEnvironment( hooks, testEnvironment, "afterEach" );
+		setHookFromEnvironment( hooks, testEnvironment, "after" );
+
+		function setHookFromEnvironment( hooks, environment, name ) {
+			const potentialHook = environment[ name ];
+			hooks[ name ] = typeof potentialHook === "function" ? [ potentialHook ] : [];
```

Comment:
I'm unsure if it matters, but this creates 4 new arrays per module regardless of if they are needed.  Instead of falling back to `[]` we could set to `null`, and in `setHookFunction` below you could check for null and create the array when needed.